File: February/feb14/feb14.js
class Solution {
    constructor() {
        this.first = null;
        this.second = null;
        this.prev = null;
    }

    inorder(root) {
        if (!root) return;

        this.inorder(root.left); // Traverse left subtree

        // Identify swapped nodes
        if (this.prev && root.val < this.prev.val) {
            if (!this.first) this.first = this.prev; // First incorrect node
            this.second = root;  // Second incorrect node (updated if needed)
        }

        this.prev = root; // Move to the next node

        this.inorder(root.right); // Traverse right subtree
    }

    correctBST(root) {
        this.inorder(root); // Identify misplaced nodes
        if (this.first && this.second) {
            [this.first.val, this.second.val] = [this.second.val, this.first.val]; // Swap values to restore BST
        }
    }
}

File: February/feb14/feb14.cpp
class Solution {
private:
    Node *first = nullptr, *second = nullptr, *prev = nullptr;

    // Inorder traversal to find the misplaced nodes
    void inorder(Node* root) {
        if (!root) return; // Base case: If node is null, return

        inorder(root->left); // Traverse the left subtree

        // Identify swapped nodes
        if (prev && root->data < prev->data) {  
            if (!first) first = prev; // First incorrect node (first occurrence)
            second = root;  // Second incorrect node (could be updated)
        }
        
        prev = root; // Move to the next node
        
        inorder(root->right); // Traverse the right subtree
    }

public:
    // Function to correct the swapped nodes in BST
    void correctBST(Node* root) {
        inorder(root); // Perform inorder traversal to detect misplaced nodes
        if (first && second) std::swap(first->data, second->data); // Swap their values to restore BST property
    }
};
